File: Python3/0034_find_first_and_last_position_of_element_in_sorted_array.py
# https://leetcode-cn.com/problems/find-first-and-last-position-of-element-in-sorted-array/

class Solution:
    def searchRange(self, nums, target):
        """
        Inputs:
            nums: list[int]
            target: int
        Outputs:
            res: list[int]
        """

        if nums == []:
            return [-1, -1]

        l = self.find_first(nums, target)
        if l == -1:
            r = -1
        else:
            r = self.find_last(nums, target)

        res = [l, r]
        return res

    def find_first(self, nums, target):
        """
        Inputs:
            nums: list[int]
            target: int
        Outputs:
            res: int
        """

        # Find the first number that is greater than or equal to target.
        # Return its index.

        l = 0
        r = len(nums) - 1
        while l < r:
            mid = (l + r) // 2
            if nums[mid] < target:
                l = mid + 1
            else:
                r = mid

        # Check if this number is equal to target.
        if nums[l] != target:
            l = -1

        res = l
        return res

    def find_last(self, nums, target):
        """
        Inputs:
            nums: list[int]
            target: int
        Outputs:
            res: int
        """

        # Find the first number that is greater than target.
        # Return the index of the last number before this one.

        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] <= target:
                l = mid + 1
            else:
                r = mid - 1

        res = l - 1
        return res

File: Cpp/0034_find_first_and_last_position_of_element_in_sorted_array.cpp
// https://leetcode.cn/problems/find-first-and-last-position-of-element-in-sorted-array/

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res(2);
        res[0] = searchFirst(nums, target);
        res[1] = searchLast(nums, target);
        return res;
    }

private:
    int searchFirst(vector<int> nums, int target) {
        int l = 0, r = nums.size() - 1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (nums[mid] == target) {
                if ((mid == 0) || (nums[mid-1] != target)) {
                    return mid;
                }
                else {
                    r = mid - 1;
                }
            }
            else if (nums[mid] < target) l = mid + 1;
            else r = mid - 1;
        }
        return -1;
    }

    int searchLast(vector<int> nums, int target) {
        int l = 0, r = nums.size() - 1;
        while (l <= r) {
            int mid = (l + r) / 2;
            if (nums[mid] == target) {
                if ((mid == nums.size()-1) || (nums[mid+1] != target)) {
                    return mid;
                }
                else {
                    l = mid + 1;
                }
            }
            else if (nums[mid] < target) l = mid + 1;
            else r = mid - 1;
        }
        return -1;
    }
};